### ISSUE TITLE
Email / Password possible?

### ISSUE THREAD

**Issue by juliusspencer** [OpenFormTech/FirebaseDeviceFlow#4]
Hi, does anyone know if it is possible to do email password Firebase auth?

**Comment by JLefebvre55** [comment]
Hey @juliusspencer

You could have a custom implementation of the OAuth2 "Device Flow" authentication process, without relying on external providers. You would essentially be a) generating a device code and login URL, b) handling device code sign-in using email and password at the login URL, and c) generating a custom token upon form completion. You'd be effectively replacing the provider in the process.

**Comment by juliusspencer** [comment]
Hi @JLefebvre55 thanks for the reply. The one piece I'm unsure of is how to convert the token into a firebase credential.
```
    firebaseCredential(tokenResponse : TokenResponse) {
        return this.firebaseProvider.credential(tokenResponse.access_token);
    }
```
I'm not sure how this part would work. Somehow I need to pass a token to firebase and have it create a credential. I'm not sure if the firebase API allows me to do that.

**Comment by JLefebvre55** [comment]
@juliusspencer this might help you: https://firebase.google.com/docs/auth/admin/create-custom-tokens

Essentially, once you get the device code sign-in page, [create a custom JWT](https://firebase.google.com/docs/auth/admin/create-custom-tokens#create_custom_tokens_using_a_third-party_jwt_library) at sign-in time, and return it to the original limited-input device the next time it pings. Then, use `signInWithCustomToken()`. Make sure you have the relevant functionality enabled in the Firebase console.

**Comment by juliusspencer** [comment]
Thanks @JLefebvre55 , I ended up making a website to enter in email / password and endpoints:
1. to create a device token and url
2. to process email / password and validate device token
3. to try to get a token based on a device token

I also stored, in firestore, the device token and its status such as expiry timestamp and whether it had been authenticated via the website. After that I was able to use the token returned in 3 with `signInWithCustomToken()`.

I also created an Android app to perform 1 and 3 and show a QR code based on the results of 1.

I'm not sure if it's the correct way to do it, but it seems to work.
